File: bench/binary-trees/cpp/binary-trees.cpp
// binary-trees — see ../c/binary-trees.c. Plain per-node new/delete.
#include <cstdio>
#include <cstdlib>

struct Node { Node *left, *right; long item; };

static Node *newNode(Node *l, Node *r, long item) {
    Node *n = new Node;
    n->left = l; n->right = r; n->item = item;
    return n;
}
static long itemCheck(Node *t) {
    if (t->left == nullptr) return t->item;
    return t->item + itemCheck(t->left) - itemCheck(t->right);
}
static Node *bottomUp(long item, int depth) {
    if (depth > 0)
        return newNode(bottomUp(2 * item - 1, depth - 1), bottomUp(2 * item, depth - 1), item);
    return newNode(nullptr, nullptr, item);
}
static void deleteTree(Node *t) {
    if (t->left != nullptr) { deleteTree(t->left); deleteTree(t->right); }
    delete t;
}
int main(int argc, char **argv) {
    int N = argc > 1 ? std::atoi(argv[1]) : 10;
    int minDepth = 4;
    int maxDepth = (minDepth + 2 > N) ? minDepth + 2 : N;
    int stretchDepth = maxDepth + 1;

    Node *stretch = bottomUp(0, stretchDepth);
    std::printf("stretch tree of depth %d\t check: %ld\n", stretchDepth, itemCheck(stretch));
    deleteTree(stretch);

    Node *longLived = bottomUp(0, maxDepth);
    for (int depth = minDepth; depth <= maxDepth; depth += 2) {
        long iterations = 1L << (maxDepth - depth + minDepth);
        long check = 0;
        for (long i = 1; i <= iterations; i++) {
            Node *t1 = bottomUp(i, depth);
            check += itemCheck(t1);
            deleteTree(t1);
            Node *t2 = bottomUp(-i, depth);
            check += itemCheck(t2);
            deleteTree(t2);
        }
        std::printf("%ld\t trees of depth %d\t check: %ld\n", iterations * 2, depth, check);
    }
    std::printf("long lived tree of depth %d\t check: %ld\n", maxDepth, itemCheck(longLived));
    deleteTree(longLived);
    return 0;
}

File: bench/binary-trees/rust/binary-trees.rs
// binary-trees — see ../c/binary-trees.c. Box per node; trees drop at scope exit.
struct Node {
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
    item: i64,
}
fn bottom_up(item: i64, depth: i32) -> Box<Node> {
    if depth > 0 {
        Box::new(Node {
            left: Some(bottom_up(2 * item - 1, depth - 1)),
            right: Some(bottom_up(2 * item, depth - 1)),
            item,
        })
    } else {
        Box::new(Node { left: None, right: None, item })
    }
}
fn item_check(t: &Node) -> i64 {
    match &t.left {
        None => t.item,
        Some(l) => t.item + item_check(l) - item_check(t.right.as_ref().unwrap()),
    }
}
fn main() {
    let n: i32 = std::env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(10);
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };
    let stretch_depth = max_depth + 1;

    let stretch = bottom_up(0, stretch_depth);
    println!("stretch tree of depth {}\t check: {}", stretch_depth, item_check(&stretch));
    drop(stretch);

    let long_lived = bottom_up(0, max_depth);
    let mut depth = min_depth;
    while depth <= max_depth {
        let iterations: i64 = 1i64 << (max_depth - depth + min_depth);
        let mut check: i64 = 0;
        for i in 1..=iterations {
            check += item_check(&bottom_up(i, depth));
            check += item_check(&bottom_up(-i, depth));
        }
        println!("{}\t trees of depth {}\t check: {}", iterations * 2, depth, check);
        depth += 2;
    }
    println!("long lived tree of depth {}\t check: {}", max_depth, item_check(&long_lived));
}
